File: services/requirement.service.ts
import { requireUserId } from '@/lib/auth-helpers';
import {
  REQUIREMENT_COLUMNS,
  REQUIREMENT_STATUS_COLUMNS,
} from '@/lib/supabase-live-schema';
import { supabase } from '@/lib/supabase';
import { DocumentType } from '@/constants/documentTypes';
import {
  REQUIREMENT_CATEGORY_LABELS,
  RequirementCategory,
} from '@/constants/requirementCategories';
import type { Document } from '@/types/document';
import type { Requirement, RequirementStatusValue } from '@/types/requirement';
import type { ExtractedRequirementsData } from '@/types/requirement';

const REQUIRED_CATEGORIES = new Set<RequirementCategory>([
  RequirementCategory.Eligibility,
  RequirementCategory.Documents,
  RequirementCategory.Language,
  RequirementCategory.Academic,
  RequirementCategory.Recommendation,
  RequirementCategory.Identity,
]);

function hasDocumentForTypes(documents: Document[], types: DocumentType[]): boolean {
  return documents.some((doc) => doc.document_type && types.includes(doc.document_type));
}

/** Derives fulfillment from uploaded documents and requirement category. */
export function evaluateRequirementFulfillment(
  category: RequirementCategory,
  documents: Document[],
): boolean {
  switch (category) {
    case RequirementCategory.Documents:
      return documents.length > 0;
    case RequirementCategory.Language:
      return hasDocumentForTypes(documents, [DocumentType.IeltsScore, DocumentType.ToeflScore]);
    case RequirementCategory.Academic:
      return hasDocumentForTypes(documents, [DocumentType.AcademicTranscript]);
    case RequirementCategory.Recommendation:
      return hasDocumentForTypes(documents, [DocumentType.LetterOfRecommendation]);
    case RequirementCategory.Identity:
      return hasDocumentForTypes(documents, [DocumentType.Passport]);
    case RequirementCategory.Eligibility:
      return hasDocumentForTypes(documents, [DocumentType.Cv, DocumentType.AcademicTranscript]);
    case RequirementCategory.Financial:
      return hasDocumentForTypes(documents, [DocumentType.Other]);
    case RequirementCategory.Deadline:
      return true;
    default:
      return false;
  }
}

function enrichRequirement(
  row: Pick<Requirement, 'id' | 'application_id' | 'category' | 'created_at'>,
  status: RequirementStatusValue | null,
): Requirement {
  const category = row.category as RequirementCategory;

  return {
    ...row,
    category,
    title: REQUIREMENT_CATEGORY_LABELS[category] ?? category,
    is_required: REQUIRED_CATEGORIES.has(category),
    is_fulfilled: status === 'fulfilled',
    description: null,
    eligibility: null,
    language_requirement: null,
    gpa_requirement: null,
    recommendation_letters_count: null,
    passport_requirement: null,
    source_url: null,
    extracted_data: null,
  };
}

async function loadStatuses(requirementIds: string[]): Promise<Map<string, RequirementStatusValue>> {
  if (requirementIds.length === 0) return new Map();

  const { data, error } = await supabase
    .from('requirement_status')
    .select(REQUIREMENT_STATUS_COLUMNS)
    .in('requirement_id', requirementIds);

  if (error) throw new Error(error.message);

  return new Map(
    (data ?? []).map((row) => [row.requirement_id, row.status as RequirementStatusValue]),
  );
}

export const RequirementService = {
  async listByApplication(applicationId: string): Promise<Requirement[]> {
    const { data, error } = await supabase
      .from('requirements')
      .select(REQUIREMENT_COLUMNS)
      .eq('application_id', applicationId)
      .order('created_at', { ascending: true });

    if (error) throw new Error(error.message);

    const rows = data ?? [];
    const statusMap = await loadStatuses(rows.map((row) => row.id));

    return rows.map((row) =>
      enrichRequirement(row as Requirement, statusMap.get(row.id) ?? null),
    );
  },

  async createFromExtractedData(
    applicationId: string,
    data: ExtractedRequirementsData,
  ): Promise<Requirement[]> {
    await requireUserId();
    const now = new Date().toISOString();

    const categories = [
      RequirementCategory.Eligibility,
      RequirementCategory.Language,
      RequirementCategory.Academic,
      RequirementCategory.Recommendation,
      RequirementCategory.Identity,
      ...data.required_documents.map(() => RequirementCategory.Documents),
    ];

    const { data: inserted, error } = await supabase
      .from('requirements')
      .insert(
        categories.map((category) => ({
          application_id: applicationId,
          category,
          created_at: now,
        })),
      )
      .select(REQUIREMENT_COLUMNS);

    if (error) throw new Error(error.message);

    return (inserted ?? []).map((row) => enrichRequirement(row as Requirement, 'missing'));
  },


  async createDefaultCategories(applicationId: string): Promise<Requirement[]> {
    await requireUserId();
    const now = new Date().toISOString();

    const categories = [
      RequirementCategory.Eligibility,
      RequirementCategory.Documents,
      RequirementCategory.Language,
      RequirementCategory.Academic,
      RequirementCategory.Recommendation,
      RequirementCategory.Identity,
    ];

    const { data: inserted, error } = await supabase
      .from('requirements')
      .insert(
        categories.map((category) => ({
          application_id: applicationId,
          category,
          created_at: now,
        })),
      )
      .select(REQUIREMENT_COLUMNS);

    if (error) throw new Error(error.message);

    return (inserted ?? []).map((row) => enrichRequirement(row as Requirement, 'missing'));
  },

  /** Updates `requirement_status` rows based on uploaded documents. */
  async syncStatusFromDocuments(
    applicationId: string,
    documents: Document[],
  ): Promise<Requirement[]> {
    await requireUserId();

    const requirements = await this.listByApplication(applicationId);
    if (requirements.length === 0) return requirements;

    const requirementIds = requirements.map((req) => req.id);
    const { data: existingStatuses, error: statusError } = await supabase
      .from('requirement_status')
      .select(REQUIREMENT_STATUS_COLUMNS)
      .in('requirement_id', requirementIds);

    if (statusError) throw new Error(statusError.message);

    const existingByRequirement = new Map(
      (existingStatuses ?? []).map((row) => [row.requirement_id, row]),
    );

    for (const requirement of requirements) {
      const fulfilled = evaluateRequirementFulfillment(requirement.category, documents);
      const status: RequirementStatusValue = fulfilled ? 'fulfilled' : 'missing';
      const existing = existingByRequirement.get(requirement.id);

      if (existing) {
        if (existing.status !== status) {
          const { error } = await supabase
            .from('requirement_status')
            .update({ status })
            .eq('id', existing.id);

          if (error) throw new Error(error.message);
        }
      } else {
        const { error } = await supabase
          .from('requirement_status')
          .insert({ requirement_id: requirement.id, status });

        if (error) throw new Error(error.message);
      }
    }

    return this.listByApplication(applicationId);
  },
};

File: backend/app/services/requirement_service.py
"""Requirement data access and fulfillment evaluation."""

from app.db.schema import REQUIREMENT_COLUMNS, REQUIREMENT_STATUS_COLUMNS
from app.db.supabase import get_supabase_client
from app.models import Document, Requirement
from app.models.constants import (
    REQUIREMENT_CATEGORY_LABELS,
    REQUIRED_CATEGORIES,
    RequirementCategory,
)
from app.services.application_service import ApplicationService


def _has_document_for_types(documents: list[Document], types: list[str]) -> bool:
    return any(document.document_type in types for document in documents)


def evaluate_requirement_fulfillment(category: str, documents: list[Document]) -> bool:
    """Mirrors frontend RequirementService.evaluateRequirementFulfillment."""
    match category:
        case RequirementCategory.DOCUMENTS:
            return len(documents) > 0
        case RequirementCategory.LANGUAGE:
            return _has_document_for_types(documents, ["ielts_score", "toefl_score"])
        case RequirementCategory.ACADEMIC:
            return _has_document_for_types(documents, ["academic_transcript"])
        case RequirementCategory.RECOMMENDATION:
            return _has_document_for_types(documents, ["letter_of_recommendation"])
        case RequirementCategory.IDENTITY:
            return _has_document_for_types(documents, ["passport"])
        case RequirementCategory.ELIGIBILITY:
            return _has_document_for_types(documents, ["cv", "academic_transcript"])
        case RequirementCategory.FINANCIAL:
            return _has_document_for_types(documents, ["other"])
        case RequirementCategory.DEADLINE:
            return True
        case _:
            return False


def _enrich_requirement(row: dict, status: str | None) -> Requirement:
    category = row["category"]
    return Requirement(
        id=row["id"],
        application_id=row["application_id"],
        category=category,
        created_at=row["created_at"],
        title=REQUIREMENT_CATEGORY_LABELS.get(category, category),
        is_required=category in REQUIRED_CATEGORIES,
        is_fulfilled=status == "fulfilled",
    )


class RequirementService:
    @staticmethod
    def _load_statuses(requirement_ids: list[str]) -> dict[str, str]:
        if not requirement_ids:
            return {}

        response = (
            get_supabase_client()
            .table("requirement_status")
            .select(REQUIREMENT_STATUS_COLUMNS)
            .in_("requirement_id", requirement_ids)
            .execute()
        )

        return {
            row["requirement_id"]: row["status"]
            for row in (response.data or [])
        }

    @staticmethod
    def list_by_application(application_id: str) -> list[Requirement]:
        ApplicationService.get_by_id(application_id)

        response = (
            get_supabase_client()
            .table("requirements")
            .select(REQUIREMENT_COLUMNS)
            .eq("application_id", application_id)
            .order("created_at")
            .execute()
        )

        rows = response.data or []
        status_map = RequirementService._load_statuses([row["id"] for row in rows])
        return [_enrich_requirement(row, status_map.get(row["id"])) for row in rows]

    @staticmethod
    def sync_status_from_documents(
        application_id: str,
        documents: list[Document],
    ) -> list[Requirement]:
        """Mirrors frontend RequirementService.syncStatusFromDocuments."""
        requirements = RequirementService.list_by_application(application_id)
        if not requirements:
            return requirements

        client = get_supabase_client()
        requirement_ids = [requirement.id for requirement in requirements]
        response = (
            client.table("requirement_status")
            .select(REQUIREMENT_STATUS_COLUMNS)
            .in_("requirement_id", requirement_ids)
            .execute()
        )
        existing_by_requirement = {
            row["requirement_id"]: row for row in (response.data or [])
        }

        for requirement in requirements:
            fulfilled = evaluate_requirement_fulfillment(requirement.category, documents)
            status = "fulfilled" if fulfilled else "missing"
            existing = existing_by_requirement.get(requirement.id)

            if existing:
                if existing["status"] != status:
                    (
                        client.table("requirement_status")
                        .update({"status": status})
                        .eq("id", existing["id"])
                        .execute()
                    )
            else:
                (
                    client.table("requirement_status")
                    .insert({"requirement_id": requirement.id, "status": status})
                    .execute()
                )

        return RequirementService.list_by_application(application_id)

    @staticmethod
    def format_for_prompt(requirements: list[Requirement]) -> str:
        if not requirements:
            return "No structured requirements recorded for this application."

        lines: list[str] = []
        for requirement in requirements:
            fulfillment = "fulfilled" if requirement.is_fulfilled else "missing"
            required = "required" if requirement.is_required else "optional"
            lines.append(
                f"- {requirement.title} ({requirement.category}): "
                f"{fulfillment}, {required}"
            )
        return "\n".join(lines)
